packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp

```diff
@@ -0,0 +1,306 @@
+        if(boundary_edges[i] < boundary_edges[i+1]){
+          edge_set.insert(std::make_pair(boundary_edges[i],boundary_edges[i+1]));
+        } else {
+          edge_set.insert(std::make_pair(boundary_edges[i+1],boundary_edges[i]));
+        }
+      }
+    }
+    //factor in the locally-owned boundary edges
+    for(int i = 0; i < num_boundary_edges*2; i++){
+      if(map->getLocalElement(boundary_edges[i]) != fail){
+        local_boundary_counts[map->getLocalElement(boundary_edges[i])]++;
+      }
+    }
+    
+    //use the received boundary edges, but check to make sure they haven't been used yet.
+    for(int i = 0; i < recvsize; i+=2){
+      //make sure the edge has not been used (if count returns 1 for one of these, the edge has been used)
+      if(edge_set.count(std::make_pair(recvbuf[i], recvbuf[i+1])) == 0 && edge_set.count(std::make_pair(recvbuf[i+1],recvbuf[i])) == 0){
+        //check which endpoints are owned locally, and set the boundary counts appropriately
+        if(map->getLocalElement(recvbuf[i]) != fail){
+          local_boundary_counts[map->getLocalElement(recvbuf[i])]++;
+        }
+        if(map->getLocalElement(recvbuf[i+1]) != fail){
+          local_boundary_counts[map->getLocalElement(recvbuf[i+1])]++;
+        }
+        //add the received edge to the set, to ensure no received duplicates are counted
+        if(recvbuf[i] < recvbuf[i+1]){
+          edge_set.insert(std::make_pair(recvbuf[i],recvbuf[i+1]));
+        } else {
+  	  edge_set.insert(std::make_pair(recvbuf[i+1],recvbuf[i]));
+        }
+      }
+    }
+    //we should now have complete local knowledge of the boundary.
+
+    //convert adjacency array to use local identifiers instead of global.
+  
+    typename map_t::local_ordinal_type* out_edges_lid = new typename map_t::local_ordinal_type[nEdge];
+    for(size_t i = 0; i < nEdge; i++){
+      out_edges_lid[i] = mapWithCopies->getLocalElement(out_edges[i]);
+    }
+    icePropGraph<typename map_t::local_ordinal_type>* g = new icePropGraph<typename map_t::local_ordinal_type>({nVtx, nEdge, out_edges_lid,out_offsets});
+    Zoltan2::iceSheetPropagation<map_t> prop(problemComm, map, mapWithCopies, g, local_boundary_counts, grounding, nVtx, nGhosts);
+    Teuchos::ArrayRCP<const IcePropVtxLabel<lno_t,gno_t>> labels = prop.propagate();
+    
+
+    //interpret the returned values
+    
+    for(size_t i = 0; i < g->n; i++){
+      IcePropVtxLabel<lno_t,gno_t> curr_node = labels[i];
+      gno_t gid = mapWithCopies->getGlobalElement(curr_node.id);
+      std::cout<<"Vertex "<<gid<<", First label="<<curr_node.first_label<<", first sender="<<curr_node.first_sender<<", first used="<<curr_node.first_used;
+      std::cout<<", second label="<<curr_node.second_label<<", second sender="<<curr_node.second_sender<<", second_used="<<curr_node.second_used<<"\n";
+      IcePropGrounding_Status gs = curr_node.getGroundingStatus();
+      if(gs == ICEPROPGS_FULL) vertex_status[i] = -2;
```

Comment:
Instead of -2, 0, -1 for status, can we define an enumerated type against which the user can test?  Something like Zoltan2::IceGrounded, Zoltan2::IceFloating, Zoltan2::IceHinged?

Reply:
I have created this enumerated type.